File: Python/152.py
// https://en.wikipedia.org/wiki/Maximum_subarray_problem

class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        backNums = nums[::-1]
        for index in range(1, len(nums)):
            nums[index] *= nums[index-1] or 1
            backNums[index] *= backNums[index-1] or 1
        return max(nums+backNums)

File: CPP/152.cpp
// https://en.wikipedia.org/wiki/Maximum_subarray_problem

class Solution {
public:
    int maxProduct(vector<int>& nums) {
        int n = nums.size(), res = nums[0], left= 0, right = 0;
        for (int i = 0; i < n; i++){
            // If there cannot be zero in an array, it can be one big n scan
            // Since there can be zero, we can divide it into subproblems; reset when there is zero
            left = (left ? left : 1) * nums[i];
            right = (right ? right : 1) * nums[n - i - 1];
            // Compare res, left, right while reset left and right to 1 whenever encounter 0
            res = max(res, max(left, right));
        }
        return res;
    }
};